<launch>

    <!-- ############################ Tool approach offsets ############################ -->
    <!-- distance in meters from the face to the tool when begining approach -->
    <arg name="dist_face_start" value="0.08"/>
    <!-- distance in meters from the face to the tool to stop -->
    <arg name="dist_face_end" value="0.00"/>
    <!-- distance in meters from the wrist to the tool tip -->
    <arg name="tool_dist" value="0.138"/>
    <!-- ################################################################################# -->

    <!-- ############################ Tool approach offsets ############################ -->
    <node pkg="tf" type="static_transform_publisher" name="optimal_head_location_frame" args="0.78 -0.28 1.1 -0.36 0 0 /base_link /optimal_head_location 100" />
    <!-- ################################################################################# -->

    <!-- ############################ EPC actionlib server ############################### -->
    <include file="$(find hrl_rfh_summer_2011)/launch/epc_move_actionlib.launch"/>
    <!-- ################################################################################# -->

    <!-- ############################ Force torque parameters ############################ -->
    <node pkg="pr2_collision_monitor" type="force_torque_monitor" name="force_torque_monitor" output="screen" respawn="true">
        <param name="force_threshold" type="double" value="0.0005" /> <!-- not used -->
        <param name="time_constant" type="double" value="0.001" />
        <param name="history" type="int" value="500" />
        <remap from="/force_topic" to="/force_torque_ft3_Vec3"/>
    </node>
    <!-- ################################################################################# -->

    <!-- ##################### Joint collision detection parameters ###################### -->
    <node pkg="pr2_collision_monitor" type="joint_coll_detect" name="r_joint_coll_detect" output="screen" >
        <param name="arm" type="str" value="r"/>
        <param name="training_mode" type="bool" value="false"/>
        <param name="significance_mode" type="bool" value="false"/>
        <rosparam param="min_errors"> [-0.04, -0.08, -0.12, -0.10, -0.15, -0.07, -0.06] </rosparam>
        <rosparam param="max_errors"> [0.04, 0.08, 0.12, 0.10, 0.15, 0.07, 0.06] </rosparam>
        <param name="filename_prefix" type="str" value="$(find hrl_rfh_summer_2011)/data/"/>
        <rosparam param="filename_list">
            [move_coarse_ik_r_data.bag,
             move_prep_pose_r_data.bag]
        </rosparam>

    </node>                 
    <!-- ################################################################################# -->

    <!-- #################### Finger collision detection parameters ###################### -->
    <node pkg="pr2_collision_monitor" type="fingertip_monitor" name="r_fingertip_monitor" output="screen" respawn="true">
        <param name="arm" type="str" value="r" />
        <param name="sensor_threshold" type="double" value="3.0" />
        <param name="time_constant" type="double" value="0.001" />
        <param name="history" type="int" value="100" />
    </node>
    <!-- ################################################################################# -->
    
    <!-- ################################### Kinematics ################################## -->
    <!--
    <include file="$(find pr2_arm_kinematics)/launch/pr2_ik_rarm_node.launch"/>
    <include file="$(find pr2_arm_kinematics)/launch/pr2_ik_larm_node.launch"/>
    -->
    <!-- ################################################################################# -->

    <!-- ################################ Approach services ############################## -->
    <!-- Checks a vx,vy,vtheta and returns the planner/costmap score -->
    <node pkg="costmap_services" name="costmap_services" type="costmap_services" 
          ns="pr2_approach_table" respawn="false" output="screen">
        <rosparam file="$(find pr2_approach_table)/launch/costmap.yaml" command="load" />
    </node>
    <!-- Performs approach -->
    <node pkg="pr2_approach_table" name="approach_services" type="approach_services.py" 
          args="--ns /pr2_approach_table" respawn="false" output="screen">
        <remap from="approach_cmd_vel" to="/navigation/cmd_vel" />
    </node>
    <!-- Rotate / Backup Node  -->
    <node pkg="rfid_behaviors" name="rotate_backup_node" 
	  type="rotate_backup_node.py" output="screen" respawn="false">
    </node>
  <!-- UNFOLD: rosservice call /traj_playback/unfold 0 -->
    <!-- The IK are already listed in rfid_services.launch -->
    <node pkg="hrl_trajectory_playback" name="unfold_traj_playback" type="playback.py" 
	  output="screen" respawn="false" 
	  args="--pkl $(find hrl_rfh_summer_2011)/data/r_slow_untuck.pkl --name unfold" />
    <!-- ################################################################################# -->
    
    <!-- ############################## Navigation Stack ################################# -->
    <include file="$(find hrl_rfh_summer_2011)/launch/willow_nav.launch" />
    <!-- ################################################################################# -->

    <!-- ##################### Tool static transforms (do not touch!) #################### -->
    <node pkg="tf" type="static_transform_publisher" name="l_scratcher_tip_frame" args="$(arg tool_dist) 0 0 0 0 0 /l_gripper_tool_frame /l_scratcher_tip 100" />
    <node pkg="tf" type="static_transform_publisher" name="l_scratcher_touch_frame" args="-$(arg dist_face_end) 0 0 0 0 0 /l_scratcher_tip /l_scratcher_touch 100" />
    <node pkg="tf" type="static_transform_publisher" name="l_scratcher_approach_frame" args="$(arg dist_face_start) 0 0 0 0 0 /l_scratcher_tip /l_scratcher_approach 100" />

    <node pkg="tf" type="static_transform_publisher" name="r_scratcher_tip_frame" args="$(arg tool_dist) 0 0 0 0 0 /r_gripper_tool_frame /r_scratcher_tip 100" />
    <node pkg="tf" type="static_transform_publisher" name="r_scratcher_touch_frame" args="-$(arg dist_face_end) 0 0 0 0 0 /r_scratcher_tip /r_scratcher_touch 100" />
    <node pkg="tf" type="static_transform_publisher" name="r_scratcher_approach_frame" args="$(arg dist_face_start) 0 0 0 0 0 /r_scratcher_tip /r_scratcher_approach 100" />
    <!-- ################################################################################# -->

</launch>
